File: spatc/spatc/Models/TaiKhoan.cs
using System.ComponentModel.DataAnnotations;

namespace spatc.Models
{
    public class TaiKhoan
    {
        [Key]
        public int Id { get; set; }

        [Required(ErrorMessage = "Email không được để trống")]
        [Display(Name = "Email")]
        [StringLength(100)]
        [EmailAddress(ErrorMessage = "Email không hợp lệ")]
        public string Email { get; set; } = string.Empty;

        [Required(ErrorMessage = "Mật khẩu không được để trống")]
        [Display(Name = "Mật khẩu")]
        [StringLength(255)]
        public string MatKhau { get; set; } = string.Empty; // Lưu dạng hash

        [Display(Name = "Họ tên")]
        [StringLength(100)]
        public string? HoTen { get; set; }

        [Display(Name = "Vai trò")]
        [StringLength(50)]
        public string VaiTro { get; set; } = "User"; // Admin, NhanVien (Staff), User

        [Display(Name = "Ngày tạo")]
        public DateTime NgayTao { get; set; } = DateTime.Now;

        [Display(Name = "Trạng thái")]
        public bool TrangThai { get; set; } = true; // true = hoạt động, false = khóa

        [Display(Name = "Provider")]
        [StringLength(50)]
        public string? Provider { get; set; } // "Facebook", "Google", null = local

        [Display(Name = "Provider ID")]
        [StringLength(255)]
        public string? ProviderId { get; set; } // ID từ provider

        // Navigation property: Một tài khoản có thể có một khách hàng
        public virtual KhachHang? KhachHang { get; set; }
    }
}

File: api/schemas/TaiKhoan.js
const mongoose = require("mongoose");

const taiKhoanSchema = new mongoose.Schema(
  {
    email: {
      type: String,
      required: [true, "Email is required"],
      unique: true,
      lowercase: true,
      trim: true,
      match: [
        /^\w+([\\.-]?\w+)*@\w+([\\.-]?\w+)*(\.\w{2,3})+$/,
        "Please provide a valid email",
      ],
    },

    matKhau: {
      type: String,
      required: [true, "Password is required"],
      minlength: 6,
      select: false,
    },

    hoTen: {
      type: String,
      trim: true,
      maxlength: [100, "Name cannot exceed 100 characters"],
    },

    vaiTro: {
      type: String,
      enum: {
        values: ["user", "staff", "admin"],
        message: "Role must be user, staff, or admin",
      },
      default: "user",
    },

    trangThai: {
      type: Boolean,
      default: true,
    },

    provider: {
      type: String,
      enum: ["facebook", "google", null],
      default: null,
    },

    providerId: {
      type: String,
      sparse: true,
    },

    ngayTao: {
      type: Date,
      default: Date.now,
    },

    ngayCapNhat: {
      type: Date,
      default: Date.now,
    },

    lastLogin: {
      type: Date,
      default: null,
    },

    emailVerified: {
      type: Boolean,
      default: false,
    },

    emailVerificationToken: {
      type: String,
      default: null,
    },

    passwordResetToken: {
      type: String,
      default: null,
    },

    passwordResetExpires: {
      type: Date,
      default: null,
    },
  },
  {
    timestamps: true,
  },
);

taiKhoanSchema.index({ email: 1 });
taiKhoanSchema.index({ providerId: 1 });
taiKhoanSchema.index({ vaiTro: 1 });

taiKhoanSchema.methods.toJSON = function () {
  const obj = this.toObject();
  delete obj.matKhau;
  delete obj.passwordResetToken;
  delete obj.emailVerificationToken;
  return obj;
};

taiKhoanSchema.pre("save", function (next) {
  this.ngayCapNhat = Date.now();
  next();
});

module.exports = mongoose.model("TaiKhoan", taiKhoanSchema);
